velox/functions/lib/DateTimeFormatter.h

```diff
@@ -0,0 +1,403 @@
+  DAY_OF_WEEK_1_BASED = 6,
+
+  // Day of week, e.g: "Tuesday" or "Tue", depending on number of times provided
+  // formatter character repeats
+  DAY_OF_WEEK_TEXT = 7,
+
+  // Year, e.g: 1996
+  YEAR = 8,
+
+  // Day of year, e.g: 189
+  DAY_OF_YEAR = 9,
+
+  // Month of year, e.g: 07, or 7 depending on number of times provided
+  // formatter character repeats
+  MONTH_OF_YEAR = 10,
+
+  // Month of year, e.g. Dec, December depending on number of times provided
+  // formatter character repeats
+  MONTH_OF_YEAR_TEXT = 11,
+
+  // Day of month, e.g: 10, 01, 001, with/without padding 0s depending on number
+  // of times provided formatter character repeats
+  DAY_OF_MONTH = 12,
+
+  // Halfday of day, e.g: "PM"
+  HALFDAY_OF_DAY = 13,
+
+  // Hour of halfday (0~11)
+  HOUR_OF_HALFDAY = 14,
+
+  // Clockhour of halfday (1~12)
+  CLOCK_HOUR_OF_HALFDAY = 15,
+
+  // Hour of day (0~23)
+  HOUR_OF_DAY = 16,
+
+  // Clockhour of day (1~24)
+  CLOCK_HOUR_OF_DAY = 17,
+
+  // Minute of hour, e.g: 30
+  MINUTE_OF_HOUR = 18,
+
+  // Second of minute, e.g: 55
+  SECOND_OF_MINUTE = 19,
+
+  // Fraction of second, e.g: 978
+  FRACTION_OF_SECOND = 20,
+
+  // Timezone, e.g: "Pacific Standard Time" or "PST"
+  TIMEZONE = 21,
+
+  // Timezone offset/id, e.g: "-0800", "-08:00" or "America/Los_Angeles"
+  TIMEZONE_OFFSET_ID = 22,
+
+  // A literal % character
+  LITERAL_PERCENT = 23
+};
+
+struct FormatPattern {
+  DateTimeFormatSpecifier specifier;
+  size_t minRepresentDigits;
+};
+
+struct DateTimeToken {
+  enum class Type { kPattern, kLiteral } type;
+  union {
+    FormatPattern pattern;
+    std::string_view literal;
+  };
+
+  explicit DateTimeToken(const FormatPattern& pattern)
+      : type(Type::kPattern), pattern(pattern) {}
+
+  explicit DateTimeToken(const std::string_view& literal)
+      : type(Type::kLiteral), literal(literal) {}
+
+  bool operator==(const DateTimeToken& right) const {
+    if (type == right.type) {
+      if (type == Type::kLiteral) {
+        return literal == right.literal;
+      } else {
+        return pattern.specifier == right.pattern.specifier &&
+            pattern.minRepresentDigits == right.pattern.minRepresentDigits;
+      }
+    }
+    return false;
+  }
+};
+
+struct DateTimeResult {
+  Timestamp timestamp;
+  int64_t timezoneId{-1};
+};
+
+class DateTimeFormatter {
+ public:
+  explicit DateTimeFormatter(
+      std::unique_ptr<char[]>&& literalBuf,
+      size_t bufSize,
+      std::vector<DateTimeToken>&& tokens)
+      : literalBuf_(std::move(literalBuf)),
+        bufSize_(bufSize),
+        tokens_(std::move(tokens)) {}
+
+  const std::unique_ptr<char[]>& literalBuf() const {
+    return literalBuf_;
+  }
+
+  size_t bufSize() const {
+    return bufSize_;
+  }
+
+  const std::vector<DateTimeToken>& tokens() const {
+    return tokens_;
+  }
+
+ private:
+  std::unique_ptr<char[]> literalBuf_;
+  size_t bufSize_;
+  std::vector<DateTimeToken> tokens_;
+};
+
+class DateTimeFormatterBuilder {
```

Comment:
Got to ask: "what exactly this class would be building?"
A format string?
So later we can turn a date/time to a string representation?
Would be nice to have a brief explanation and an example of what we are building and why do we need that (what we have built).

Reply:
Probably more proper to put above DateTimeFormatter instead of builder. Then builder will be self explanatory.